modules/fb_instant_articles_display/fb_instant_articles_display.module

```diff
@@ -5,6 +5,134 @@
+    // @see fb_instant_articles_display_module_implements_alter()
+    $wrapper = \Drupal\fb_instant_articles\ArticleWrapper::create();
+    $wrapper->entity_type = $type;
+    $wrapper->entity = $entity;
+
+    $entity->fb_instant_article_wrapper = $wrapper;
+  }
+}
+
+/**
+ * Implements hook_field_formatter_view().
+ */
+function fb_instant_articles_display_field_formatter_view($entity_type, $entity, $field, $instance, $langcode, $items, $display) {
+  // Grab the InstantArticle object we're going to be building up. If it's not
+  // present, it means that a site admin selected an FB IA field formatter for
+  // a non-FB IA content type.
+  if (!isset($entity->fb_instant_article_wrapper)) {
+    return FALSE;
+  }
+
+  $article = $entity->fb_instant_article_wrapper->getArticle();
+  $entity_info = entity_get_info($entity_type);
+  $bundle_key = !empty($entity_info['entity keys']['bundle']) ? $entity_info['entity keys']['bundle'] : $entity_type;
+  $layout_settings = fb_instant_articles_display_get_layout_settings($entity_type, $entity->{$bundle_key});
+  $display_wrapper = \Drupal\fb_instant_articles_display\EntityWrapper::create($entity_type, $entity, $layout_settings, $article);
+  $display_wrapper->fieldFormatterView($field, $instance, $langcode, $items, $display);
+}
+
+/**
+ * Implements hook_module_implements_alter().
+ *
+ * @see fb_instant_articles_display_declare_entity_preprocess_hooks()
+ */
+function fb_instant_articles_display_module_implements_alter(&$implementations, $hook) {
+  // Dynamically declare hook_preprocess_ENTITY() for each Entity type. This may
+  // be as good a place as any to call this function.
+  // @see module_implements()
+  fb_instant_articles_display_declare_entity_preprocess_hooks();
+}
+
+/**
+ * Dynamically declare hook_preprocess_ENTITY() for each Entity type.
+ *
+ * This adds a theme function to control the output of entity_view(), for every
+ * entity type, when the view mode is 'fb_instant_article'. See
+ * ctools_features_declare_functions() for precedent on dynamically declared
+ * functions.
+ *
+ * See @link https://www.drupal.org/node/1443202#comment-5787770 this d.o issue comment about updating documention for hook_preprocess_ENTITY() in D8. @endlink
+ *
+ * @see hook_preprocess_HOOK()
+ * @see fb_instant_articles_display_module_implements_alter()
+ * @see fb_instant_articles_display_theme()
+ * @see theme_entity__fb_instant_article()
+ */
+function fb_instant_articles_display_declare_entity_preprocess_hooks() {
+  foreach (array_keys(entity_get_info()) as $entity_type) {
+    $code = '';
+    if (!function_exists("fb_instant_articles_display_preprocess_{$entity_type}")) {
+      $code .= 'function fb_instant_articles_display_preprocess_' . $entity_type . "(&" . '$vars' . ") {
+      if (" . '$vars' . "['view_mode'] === 'fb_instant_article') {
+        " . '$vars' . "['theme_hook_suggestions'][] = 'entity__fb_instant_article';
+      }
+    }";
+    }
+    eval($code);
```

Comment:
@m4olivei I would love that. But `hook_preprocess()` docs say:
> It is called for all theme hooks implemented as templates, **but not for theme hooks implemented as functions**. hook_preprocess_HOOK() can be used to preprocess variables for a specific theme hook, whether implemented as a template or function.

Honestly, I was wondering if we ought to skip this theme hijacking altogether, and add Views support for `Entity rendered as Instant Article`, rather than the default rendered entity option, and create a handler that just runs `$entity->fb_instant_article_wrapper->getArticle()->render('<!doctype html>', TRUE);` (it would let us remove a lot of code in the existing module).